### ISSUE TITLE
Timers skipped during daylight saving time -> standard time transition

### ISSUE THREAD

**Issue by OhNoMoreGit** [systemd/systemd#8647]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> systemd 238

### Used distribution

> Fedora

### Description

This is similar to issue #5595, however this issue is instead about the transition _away_ from daylight saving time.

On the morning of April 1st, 2018, the Australia/Sydney timezone transitioned from daylight saving time (AEDT, +1100) to standard time (AEST, +1000):

    Sat Mar 31 15:59:58 UTC 2018 == Sun Apr  1 02:59:58 AEDT 2018
    Sat Mar 31 15:59:59 UTC 2018 == Sun Apr  1 02:59:59 AEDT 2018
    Sat Mar 31 16:00:00 UTC 2018 == Sun Apr  1 02:00:00 AEST 2018
    Sat Mar 31 16:00:01 UTC 2018 == Sun Apr  1 02:00:01 AEST 2018

A systemd timer with, say:

    OnCalendar=minutely    # equivalently: OnCalendar=*:*:00

is expected to run once a minute, however it _does not run at all_ in the "duplicate" hour.

Given the ambiguity in matching calendar times to actual timestamps, it seems to me that the clearest semantics might be something like this:

* If a time component is given as a range of values (or `*`) it should match all applicable calendar times, even if they have the same numeric value for that component.
* If a time component is given as a single value, it should match just the first applicable calendar time.
* Repetitions and multiple values for time components work as normal.

Some examples (where `02:30` exists under both daylight saving time and standard time):

    OnCalendar=minutely                  # always once a minute, even across DST transition
    OnCalendar=02:30:00                  # always the first 2:30 am, never the second
    OnCalendar=00-04:30:00               # six times: 00:30, 01:30, 02:30 (first time),
                                         #     02:30 (second time), 03:30, 04:30
    OnCalendar=00,01,02,03,04:30:00      # five times: 00:30, 01:30, 02:30 (first time only),
                                         #     03:30, 04:30
    OnCalendar=00-04/2:30:00             # three times: 00:30, 02:30 (first time only), 03:30

I was unable to wrangle the `calendarspec.c` code into something that could actually do this (there needs to be some way to ask libc things like "how many seconds are there from this timestamp until the next midnight?"), but here's a few test cases that might be useful to validate the desired behaviour. Note that only the two tests that result in a time within the duplicate hour are currently failing.

    diff --git a/src/test/test-calendarspec.c b/src/test/test-calendarspec.c
    index cce305ed2b..a372782ac1 100644
    --- a/src/test/test-calendarspec.c
    +++ b/src/test/test-calendarspec.c
    @@ -233,6 +233,27 @@ int main(int argc, char* argv[]) {
             test_next("2017-09-09 20:42:00 Pacific/Auckland", "", 12345, 1504946520000000);
             test_next("2017-09-09 20:42:00 Pacific/Auckland", "EET", 12345, 1504946520000000);

    +        // Sun Apr  1 01:59:59 AEDT 2018 → Sun Apr  1 02:15:00 AEDT 2018
    +        test_next("02:15:00", "Australia/Sydney", 1522508399999999, 1522509300000000);
    +        // Sun Apr  1 02:15:00 AEDT 2018 → Sun Apr  2 02:15:00 AEST 2018
    +        test_next("02:15:00", "Australia/Sydney", 1522509300000000, 1522599300000000);
    +
    +        // Sun Apr  1 01:59:59 AEDT 2018 → Sun Apr  1 02:30:00 AEDT 2018
    +        test_next("*:30:00", "Australia/Sydney", 1522508399999999, 1522510200000000);
    +        // Sun Apr  1 02:30:00 AEDT 2018 → Sun Apr  1 02:30:00 AEST 2018
    +        test_next("*:30:00", "Australia/Sydney", 1522510200000000, 1522513800000000);  // CURRENTLY FAILS
    +        // Sun Apr  1 02:30:00 AEST 2018 → Sun Apr  1 03:30:00 AEST 2018
    +        test_next("*:30:00", "Australia/Sydney", 1522513800000000, 1522517400000000);
    +
    +        // Sun Apr  1 01:59:59 AEDT 2018 → Sun Apr  1 02:00:00 AEDT 2018
    +        test_next("minutely", "Australia/Sydney", 1522508399999999, 1522508400000000);
    +        // Sun Apr  1 02:00:00 AEDT 2018 → Sun Apr  1 02:01:00 AEDT 2018
    +        test_next("minutely", "Australia/Sydney", 1522508400000000, 1522508460000000);
    +        // Sun Apr  1 02:59:00 AEDT 2018 → Sun Apr  1 02:00:00 AEST 2018
    +        test_next("minutely", "Australia/Sydney", 1522511940000000, 1522512000000000);  // CURRENTLY FAILS
    +        // Sun Apr  1 02:00:00 AEST 2018 → Sun Apr  1 02:01:00 AEST 2018
    +        test_next("minutely", "Australia/Sydney", 1522512000000000, 1522512060000000);
    +
             assert_se(calendar_spec_from_string("test", &c) < 0);
             assert_se(calendar_spec_from_string(" utc", &c) < 0);
             assert_se(calendar_spec_from_string("    ", &c) < 0);

**Comment by poettering** [comment]
to fix this properly I figure we should invoke mktime() twice each time when we look for the next elapsation time, one with the `.tm_isdst` set to 0 an once with it set to `.tm_isdst` to 1, and then look for which one of the two returns an earlier time.

**Comment by poettering** [comment]
I figure this is a duplicate of #5595. Closing.

**Comment by OhNoMoreGit** [comment]
Well, it isn't exactly a duplicate, as I noted in my initial comment. The bug reports are talking about different ends of daylight-saving time.

I actually do not consider #5595 to be a bug -- that is, I do not think timers that were skipped during a "missing" hour should all catch up at the end of that hour.

**Comment by zerkms** [comment]
@poettering after a rather long fresh conversation in #5595 @OhNoMoreGit could demonstrate how it's not really a duplicate: this report actually describes a case when systemd-timers behaves not as intended (skips one hour) while the other report is about a more subjective behaviour of whether systemd should or not take DST change into account at all: https://github.com/systemd/systemd/issues/5595#issuecomment-1304737859

**Comment by keszybz** [comment]
Yeah, I agree with the @OhNoMoreGit. `@minutely` should fire every minute of real time. If we don't then it's bug.

**Comment by mateon1** [comment]
I have encountered this during today's CEST -> CET timezone transition, systemd completely failed to run an `OnCalendar=*-*-* *:*:00` timer for the entire duplicated hour. I believe the behavior specified in the original issue comment is the expected one and that the current behavior is buggy.

**Comment by mbiebl** [comment]
We have a related downstream bug report: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1033548
```
02:00 timers dont run in DST transition night

aide-common ships the following timer:

[Unit]
Description=Daily AIDE check

[Timer]
OnCalendar=*-*-* 02:00:00
RandomizedDelaySec=2h
Persistent=true

[Install]
WantedBy=timers.target

This didn't run in DST transition night. I think this might be caused by
the clock jumping from 01:59 to 03:00, with 02:00 not existing.

Is there a notation to have a systemd timer run even if the exact time
the timer is supposed to run doesn't happen? I guess this might also be
the case in case of a grossly misticking clock and a timesync daemon
stepping the time, for example, from 01:59:50 to 02:02:00?

Or would be probably be a better idea to trigger a timer if systemd
finds the trigger time in the past without the timer having been
triggered?

Not running at all came as kind of surprise for me.

I might be holding things wrong but I'd like your opinion.
```

**Comment by keszybz** [comment]
It's too late in the release cycle to change things in this area. Let's try to do better for v257.
